x-pack/plugin/security/cli/src/main/java/org/elasticsearch/xpack/security/cli/AutoConfigInitialNode.java

```diff
@@ -0,0 +1,528 @@
+        // if it is, silently skip auto configuration
+        try {
+            checkExistingConfiguration(env, terminal);
+        } catch (UserException e) {
+            if (options.has(strictOption)) {
+                throw e;
+            } else {
+                return; // silent error because we wish the node to start as usual (skip auto config) if certain configurations are set
+            }
+        }
+
+        final ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC);
+        final String instantAutoConfigName = "auto_config_on_" + autoConfigDate.toInstant().getEpochSecond();
+        final Path instantAutoConfigDir = env.configFile().resolve(instantAutoConfigName);
+        try {
+            // it is useful to pre-create the sub-config dir in order to check that the config dir is writable and that
+            // file owners match (this auto-configuration script should not change the owners of config files)
+            Files.createDirectory(instantAutoConfigDir);
+            // set permissions to 750 (don't rely on umask)
+            PosixFileAttributeView view = Files.getFileAttributeView(instantAutoConfigDir, PosixFileAttributeView.class);
+            if (view != null) {
+                view.setPermissions(PosixFilePermissions.fromString("rwxr-x---"));
+            }
+        } catch (Exception e) {
+            try {
+                Files.deleteIfExists(instantAutoConfigDir);
+            } catch (Exception ex) {
+                e.addSuppressed(ex);
+            }
+            // the config dir is probably read-only, which is to be expected sometimes
+            // this will show a message to the console (the first time the node starts) and auto-configuration is effectively bypassed
+            // the message will not be subsequently shown (because auto-configuration doesn't run for node restarts)
+            if (options.has(strictOption)) {
+                throw new UserException(ExitCodes.CANT_CREATE, "Could not create auto configuration directory", e);
+            } else {
+                return; // silent error because we wish the node to start as usual (skip auto config) if config dir is not writable
+            }
+        }
+
+        // the files created or replaced by the auto-config script have the same owner as the config dir
+        if (false == Files.getOwner(instantAutoConfigDir, LinkOption.NOFOLLOW_LINKS).equals(Files.getOwner(env.configFile(),
+                LinkOption.NOFOLLOW_LINKS))) {
+            Files.deleteIfExists(instantAutoConfigDir);
+            if (options.has(strictOption)) {
+                throw new UserException(ExitCodes.CONFIG, "Aborting auto configuration because it would change file owners");
+            } else {
+                return; // if a different user runs ES compared to the user that installed it, auto configuration will not run
+            }
+        }
```

Comment:
This is the only approach so that the start-up doesn't change file owners and also that it doesn't fail the node if the ES process runs under a different user than the one that installed ES (which is obviously a supported configuration).

TBH I have qualms about the file owners issue @jkakavas .
I think it's good that we try to prevent the files in the config dir to have two different owners. In effect, by enforcing it, we ask that auto-configuration be run by the install user (which might or might not be the same as the run user). I think this is conservative and good in principle.

But it doesn't bring us tangible benefits. Even if the files have different owners, and given the permissions of all the configuration files (660), they must be part of the same group. There's no risk of access denied errors. Indeed ES already possibly creates the elasticsearch.keystore file under the user that runs ES (as opposed to the one that installed). What auto-config would do, is to also create the TLS resources under the run user, as well as changing the owner of the elasticsearch.yml file from install to the run user (we use replace instead of append because it affords an atomic change operation on the elasticsearch.yml to guarantee that ES never starts with a partial configuration). Again, there's no risk of access denied errors.

I fell that the fact that auto configuration doesn't run if the run user is different from install user is right in between a feature and a bug.